File: Platform/infra/http/modes/mock/cpp/http_client_mock.cpp
#include "http_client_mock.h"
#include <regex>
#include <thread>

namespace coyote {
namespace infra {
namespace mocks {

// RequestMatcher implementation
bool RequestMatcher::Matches(const HttpRequest& request) const {
  const auto& mock_request = static_cast<const MockHttpRequest&>(request);
  
  // Check method
  if (method != mock_request.GetMethod()) {
    return false;
  }
  
  // Check URL pattern
  if (!url_pattern.empty()) {
    std::regex pattern(url_pattern);
    if (!std::regex_match(mock_request.GetUrl(), pattern)) {
      return false;
    }
  }
  
  // Check required headers
  for (const auto& [key, value] : required_headers) {
    const auto& request_headers = mock_request.GetHeaders();
    auto it = request_headers.find(key);
    if (it == request_headers.end() || it->second != value) {
      return false;
    }
  }
  
  // Check body pattern
  if (!body_pattern.empty()) {
    std::regex pattern(body_pattern);
    if (!std::regex_match(mock_request.GetBody(), pattern)) {
      return false;
    }
  }
  
  return true;
}

// HttpClientMock implementation
HttpClientMock::HttpClientMock() {
  // Set default successful response
  default_response_ = std::make_unique<MockHttpResponse>(200, "OK");
}

std::unique_ptr<HttpResponse> HttpClientMock::Execute(const HttpRequest& request) {
  std::lock_guard<std::mutex> lock(request_mutex_);
  
  // Record the request if enabled
  if (record_requests_) {
    RecordRequest(request);
  }
  
  // Simulate latency if configured
  if (simulated_latency_.count() > 0) {
    std::this_thread::sleep_for(simulated_latency_);
  }
  
  // Simulate network error
  if (simulate_network_error_) {
    return std::make_unique<MockHttpResponse>(0, "", std::unordered_map<std::string, std::string>{}, network_error_message_);
  }
  
  // Check if we should simulate failure
  if (ShouldSimulateFailure()) {
    return std::make_unique<MockHttpResponse>(0, "", std::unordered_map<std::string, std::string>{}, "Simulated network failure");
  }
  
  // Get response from queue if available
  if (!response_queue_.empty()) {
    auto response = std::move(response_queue_.front());
    response_queue_.pop();
    return std::make_unique<MockHttpResponse>(
      response->GetStatusCode(),
      response->GetBody(),
      response->GetHeaders(),
      response->GetErrorMessage()
    );
  }
  
  // Return default response
  return std::make_unique<MockHttpResponse>(
    default_response_->GetStatusCode(),
    default_response_->GetBody(),
    default_response_->GetHeaders(),
    default_response_->GetErrorMessage()
  );
}

std::unique_ptr<HttpResponse> HttpClientMock::Get(
    const std::string& url, 
    const std::unordered_map<std::string, std::string>& headers) {
  MockHttpRequest request;
  request.SetUrl(url);
  request.SetMethod(HttpMethod::kGet);
  request.SetHeaders(headers);
  return Execute(request);
}

std::unique_ptr<HttpResponse> HttpClientMock::Post(
    const std::string& url, 
    const std::string& body,
    const std::unordered_map<std::string, std::string>& headers) {
  MockHttpRequest request;
  request.SetUrl(url);
  request.SetMethod(HttpMethod::kPost);
  request.SetBody(body);
  request.SetHeaders(headers);
  return Execute(request);
}

std::unique_ptr<HttpResponse> HttpClientMock::Put(
    const std::string& url, 
    const std::string& body,
    const std::unordered_map<std::string, std::string>& headers) {
  MockHttpRequest request;
  request.SetUrl(url);
  request.SetMethod(HttpMethod::kPut);
  request.SetBody(body);
  request.SetHeaders(headers);
  return Execute(request);
}

std::unique_ptr<HttpResponse> HttpClientMock::Delete(
    const std::string& url,
    const std::unordered_map<std::string, std::string>& headers) {
  MockHttpRequest request;
  request.SetUrl(url);
  request.SetMethod(HttpMethod::kDelete);
  request.SetHeaders(headers);
  return Execute(request);
}

void HttpClientMock::SetDefaultTimeout(long timeout_ms) {
  default_timeout_ = timeout_ms;
}

void HttpClientMock::SetDefaultHeaders(
    const std::unordered_map<std::string, std::string>& headers) {
  default_headers_ = headers;
}

void HttpClientMock::SetClientCertificate(
    const std::string& cert_path, 
    const std::string& key_path) {
  default_client_cert_path_ = cert_path;
  default_client_key_path_ = key_path;
}

void HttpClientMock::SetCACertificate(const std::string& ca_path) {
  default_ca_cert_path_ = ca_path;
}

void HttpClientMock::SetVerifyPeer(bool verify) {
  default_verify_peer_ = verify;
}

bool HttpClientMock::Ping(const std::string& url) {
  auto response = Get(url + "/health");
  return response && response->IsSuccess();
}

// Mock-specific methods
void HttpClientMock::AddResponse(std::unique_ptr<MockHttpResponse> response) {
  std::lock_guard<std::mutex> lock(request_mutex_);
  response_queue_.push(std::move(response));
}

void HttpClientMock::AddResponses(std::vector<std::unique_ptr<MockHttpResponse>> responses) {
  std::lock_guard<std::mutex> lock(request_mutex_);
  for (auto& response : responses) {
    response_queue_.push(std::move(response));
  }
}

void HttpClientMock::SetDefaultResponse(std::unique_ptr<MockHttpResponse> response) {
  std::lock_guard<std::mutex> lock(request_mutex_);
  default_response_ = std::move(response);
}

void HttpClientMock::ClearResponses() {
  std::lock_guard<std::mutex> lock(request_mutex_);
  std::queue<std::unique_ptr<MockHttpResponse>> empty;
  response_queue_.swap(empty);
}

// Request recording
void HttpClientMock::EnableRequestRecording(bool enable) {
  record_requests_ = enable;
}

const std::vector<std::unique_ptr<MockHttpRequest>>& HttpClientMock::GetRecordedRequests() const {
  return recorded_requests_;
}

void HttpClientMock::ClearRecordedRequests() {
  std::lock_guard<std::mutex> lock(request_mutex_);
  recorded_requests_.clear();
}

// Error simulation
void HttpClientMock::SimulateNetworkError(bool simulate, const std::string& error_message) {
  simulate_network_error_ = simulate;
  network_error_message_ = error_message;
}

void HttpClientMock::SetLatencySimulation(std::chrono::milliseconds latency) {
  simulated_latency_ = latency;
}

void HttpClientMock::SetFailureRate(double rate) {
  failure_rate_ = rate;
}

// Convenience methods
void HttpClientMock::AddJsonResponse(int status_code, const std::string& json_body) {
  auto response = std::make_unique<MockHttpResponse>(status_code, json_body);
  response->SetHeader("Content-Type", "application/json");
  AddResponse(std::move(response));
}

void HttpClientMock::AddSuccessResponse(const std::string& body) {
  AddResponse(std::make_unique<MockHttpResponse>(200, body));
}

void HttpClientMock::AddErrorResponse(int status_code, const std::string& error_body) {
  AddResponse(std::make_unique<MockHttpResponse>(status_code, error_body));
}

void HttpClientMock::AddNotFoundResponse() {
  AddResponse(std::make_unique<MockHttpResponse>(404, "Not Found"));
}

void HttpClientMock::AddServerErrorResponse() {
  AddResponse(std::make_unique<MockHttpResponse>(500, "Internal Server Error"));
}

void HttpClientMock::AddUnauthorizedResponse() {
  AddResponse(std::make_unique<MockHttpResponse>(401, "Unauthorized"));
}

// State inspection
size_t HttpClientMock::GetQueuedResponseCount() const {
  std::lock_guard<std::mutex> lock(request_mutex_);
  return response_queue_.size();
}

// Private helper methods
std::unique_ptr<MockHttpResponse> HttpClientMock::CreateMockResponse(
    int status_code, 
    const std::string& body, 
    const std::unordered_map<std::string, std::string>& headers) {
  return std::make_unique<MockHttpResponse>(status_code, body, headers);
}

void HttpClientMock::RecordRequest(const HttpRequest& request) {
  auto recorded_request = std::make_unique<MockHttpRequest>();
  const auto& mock_request = static_cast<const MockHttpRequest&>(request);
  
  recorded_request->SetUrl(mock_request.GetUrl());
  recorded_request->SetMethod(mock_request.GetMethod());
  recorded_request->SetHeaders(mock_request.GetHeaders());
  recorded_request->SetBody(mock_request.GetBody());
  recorded_request->SetTimeout(mock_request.GetTimeout());
  
  recorded_requests_.push_back(std::move(recorded_request));
}

bool HttpClientMock::ShouldSimulateFailure() const {
  return failure_rate_ > 0.0 && static_cast<double>(rand()) / RAND_MAX < failure_rate_;
}

}  // namespace mocks
}  // namespace infra
}  // namespace coyote

File: Platform/infra/http/modes/mock/python/http_client_mock.py
"""
Mock HTTP Client Implementation

This module provides a mock HTTP client implementation for testing,
mirroring the C++ mock implementation architecture with Python idioms.
"""

import json
import time
import random
import threading
import os
import sys
from collections import deque
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, field

# Add the parent directory to the path for imports
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', '..'))

from interfaces.python.http_client import HttpClient, HttpRequest, HttpResponse, HttpMethod


@dataclass
class RequestMatcher:
    """Matcher for setting up conditional mock responses."""
    url_pattern: str = ""
    method: Optional[HttpMethod] = None
    required_headers: Dict[str, str] = field(default_factory=dict)
    body_pattern: str = ""
    
    def matches(self, request: HttpRequest) -> bool:
        """Check if this matcher matches the given request."""
        if self.url_pattern and self.url_pattern not in request.url:
            return False
        
        if self.method and self.method != request.method:
            return False
        
        for key, value in self.required_headers.items():
            if key not in request.headers or request.headers[key] != value:
                return False
        
        if self.body_pattern and self.body_pattern not in request.body:
            return False
        
        return True


class HttpClientMock(HttpClient):
    """Mock HTTP client implementation for testing."""
    
    def __init__(self):
        """Initialize the mock HTTP client."""
        self._lock = threading.Lock()
        
        # Response queue for pre-configured responses
        self._response_queue = deque()
        
        # Default response when queue is empty
        self._default_response = HttpResponse(
            status_code=200,
            body="OK",
            headers={"Content-Type": "text/plain"},
            error_message=""
        )
        
        # Configuration
        self._default_timeout_ms = 10000
        self._default_headers = {}
        self._client_cert_path = ""
        self._client_key_path = ""
        self._ca_cert_path = ""
        self._verify_peer = True
        
        # Request recording
        self._recorded_requests = []
        self._record_requests = False
        
        # Error simulation
        self._simulate_network_error = False
        self._network_error_message = "Mock network error"
        self._simulated_latency_ms = 0
        self._failure_rate = 0.0
    
    def execute(self, request: HttpRequest) -> HttpResponse:
        """Execute an HTTP request and return the response."""
        with self._lock:
            # Record the request if recording is enabled
            if self._record_requests:
                self._recorded_requests.append(self._copy_request(request))
            
            # Simulate latency if configured
            if self._simulated_latency_ms > 0:
                time.sleep(self._simulated_latency_ms / 1000.0)
            
            # Simulate network error if configured
            if self._simulate_network_error:
                return HttpResponse(
                    status_code=0,
                    body="",
                    headers={},
                    error_message=self._network_error_message
                )
            
            # Simulate random failures if configured
            if self._failure_rate > 0 and random.random() < self._failure_rate:
                return HttpResponse(
                    status_code=500,
                    body="Simulated failure",
                    headers={},
                    error_message="Simulated random failure"
                )
            
            # Get response from queue or use default
            if self._response_queue:
                return self._response_queue.popleft()
            else:
                return self._default_response
    
    def get(self, url: str, headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a GET request."""
        request = HttpRequest(
            url=url,
            method=HttpMethod.GET,
            headers=headers or {},
            timeout_ms=self._default_timeout_ms
        )
        return self.execute(request)
    
    def post(self, url: str, body: str = "", headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a POST request."""
        request = HttpRequest(
            url=url,
            method=HttpMethod.POST,
            body=body,
            headers=headers or {},
            timeout_ms=self._default_timeout_ms
        )
        return self.execute(request)
    
    def put(self, url: str, body: str = "", headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a PUT request."""
        request = HttpRequest(
            url=url,
            method=HttpMethod.PUT,
            body=body,
            headers=headers or {},
            timeout_ms=self._default_timeout_ms
        )
        return self.execute(request)
    
    def delete(self, url: str, headers: Optional[Dict[str, str]] = None) -> HttpResponse:
        """Perform a DELETE request."""
        request = HttpRequest(
            url=url,
            method=HttpMethod.DELETE,
            headers=headers or {},
            timeout_ms=self._default_timeout_ms
        )
        return self.execute(request)
    
    def set_default_timeout(self, timeout_ms: int) -> None:
        """Set the default timeout for all requests."""
        self._default_timeout_ms = timeout_ms
    
    def set_default_headers(self, headers: Dict[str, str]) -> None:
        """Set default headers for all requests."""
        self._default_headers = headers.copy()
    
    def set_client_certificate(self, cert_path: str, key_path: str) -> None:
        """Set client certificate for SSL authentication."""
        self._client_cert_path = cert_path
        self._client_key_path = key_path
    
    def set_ca_certificate(self, ca_path: str) -> None:
        """Set CA certificate for SSL verification."""
        self._ca_cert_path = ca_path
    
    def set_verify_peer(self, verify: bool) -> None:
        """Enable or disable SSL peer verification."""
        self._verify_peer = verify
    
    def ping(self, url: str) -> bool:
        """Check connectivity to a URL."""
        # Mock ping always succeeds unless network error is simulated
        return not self._simulate_network_error
    
    # Mock-specific methods for testing
    
    def add_response(self, response: HttpResponse) -> None:
        """Add a response to the queue."""
        with self._lock:
            self._response_queue.append(response)
    
    def add_responses(self, responses: List[HttpResponse]) -> None:
        """Add multiple responses to the queue."""
        with self._lock:
            self._response_queue.extend(responses)
    
    def set_default_response(self, response: HttpResponse) -> None:
        """Set the default response for when queue is empty."""
        self._default_response = response
    
    def clear_responses(self) -> None:
        """Clear all queued responses."""
        with self._lock:
            self._response_queue.clear()
    
    def enable_request_recording(self, enable: bool = True) -> None:
        """Enable or disable request recording."""
        self._record_requests = enable
    
    def get_recorded_requests(self) -> List[HttpRequest]:
        """Get all recorded requests."""
        return self._recorded_requests.copy()
    
    def clear_recorded_requests(self) -> None:
        """Clear all recorded requests."""
        with self._lock:
            self._recorded_requests.clear()
    
    def simulate_network_error(self, simulate: bool = True, error_message: str = "Mock network error") -> None:
        """Simulate network errors."""
        self._simulate_network_error = simulate
        self._network_error_message = error_message
    
    def set_latency_simulation(self, latency_ms: int) -> None:
        """Set simulated latency for all requests."""
        self._simulated_latency_ms = latency_ms
    
    def set_failure_rate(self, rate: float) -> None:
        """Set random failure rate (0.0 to 1.0)."""
        self._failure_rate = max(0.0, min(1.0, rate))
    
    # Convenience methods for common responses
    
    def add_json_response(self, status_code: int, json_data: Any) -> None:
        """Add a JSON response to the queue."""
        response = HttpResponse(
            status_code=status_code,
            body=json.dumps(json_data),
            headers={"Content-Type": "application/json"},
            error_message=""
        )
        self.add_response(response)
    
    def add_success_response(self, body: str = "OK") -> None:
        """Add a successful response to the queue."""
        response = HttpResponse(
            status_code=200,
            body=body,
            headers={"Content-Type": "text/plain"},
            error_message=""
        )
        self.add_response(response)
    
    def add_error_response(self, status_code: int, error_body: str = "Error") -> None:
        """Add an error response to the queue."""
        response = HttpResponse(
            status_code=status_code,
            body=error_body,
            headers={"Content-Type": "text/plain"},
            error_message=""
        )
        self.add_response(response)
    
    def add_not_found_response(self) -> None:
        """Add a 404 Not Found response to the queue."""
        self.add_error_response(404, "Not Found")
    
    def add_server_error_response(self) -> None:
        """Add a 500 Internal Server Error response to the queue."""
        self.add_error_response(500, "Internal Server Error")
    
    def add_unauthorized_response(self) -> None:
        """Add a 401 Unauthorized response to the queue."""
        self.add_error_response(401, "Unauthorized")
    
    # State inspection methods
    
    def get_queued_response_count(self) -> int:
        """Get the number of queued responses."""
        return len(self._response_queue)
    
    def is_recording_requests(self) -> bool:
        """Check if request recording is enabled."""
        return self._record_requests
    
    def is_simulating_network_error(self) -> bool:
        """Check if network error simulation is enabled."""
        return self._simulate_network_error
    
    def _copy_request(self, request: HttpRequest) -> HttpRequest:
        """Create a copy of a request for recording."""
        return HttpRequest(
            url=request.url,
            method=request.method,
            body=request.body,
            headers=request.headers.copy(),
            timeout_ms=request.timeout_ms,
            client_cert_path=request.client_cert_path,
            client_key_path=request.client_key_path,
            ca_cert_path=request.ca_cert_path,
            verify_peer=request.verify_peer,
            follow_redirects=request.follow_redirects
        )